### ISSUE TITLE
Enhancements to pretrained model loading

### ISSUE THREAD

**Issue by DavidStirling** [CSBDeep/CSBDeep#69]
We're currently working towards running pretrained models on a computing cluster and some packages (most notably StarDist) utilise CSBDeep for management of model files.

A problem we've encountered is that HPC nodes don't always have internet connectivity with which to download model files, and so aimed to solve this by pointing to a series of pre-downloaded model files. However, there is currently no support within [`csbdeep.models.pretrained.get_model_folder`](https://github.com/CSBDeep/CSBDeep/blob/deb54947a734b959681ead6bde24c6143805663f/csbdeep/models/pretrained.py#L96-L102) for passing the `cache_dir` argument to keras' `get_file` method. This forces the user to always store models within the Keras default cache directory, which itself is under the user's $HOME directory. Ideally it'd be nice to have full control over where models are fetched from, the simplest approach might be checking for an environment variable such as `CSBDEEP_CACHE`.

Another issue when trying to run in parallel is that the `get_model_folder` method is set up to re-extract the downloaded model archive upon each call (`extract=True` in the`get_file` call). This seems to create a race condition when multiple workers attempt to access the same model simultaneously. Resolving this is a little more complex because the contents of the zipped model folder aren't known beforehand, but supporting some means of indicating that the models are already extracted could be useful.

Such changes would hopefully improve the utility of CSBDeep when working with computing clusters.

We'd like to acknowledge Jacob Dore-Porrett and Todd Shelper from Griffith University for helping to discover these issues.

**Comment by uschmidt83** [comment]
Hi David,

thanks for reaching out. I'd suggest to work around these issues by instead registering your own pre-trained models (see [this](https://nbviewer.org/github/CSBDeep/CSBDeep/blob/master/examples/other/technical.ipynb#Registry-for-pretrained-models) for more details). For example:

```python
from stardist.models import StarDist2D
from csbdeep.models import register_model, register_aliases

register_model(StarDist2D, '2D_versatile_fluo_local_copy', 'file:///path/to/shared/folder/python_2D_versatile_fluo.zip', '8db40dacb5a1311b8d2c447ad934fb8a')

model = StarDist2D.from_pretrained('2D_versatile_fluo_local_copy')
...
```

(You can find the links to all pre-trained models [here in the code](https://github.com/stardist/stardist/blob/41de1eb4280da307cec7dce9f081bc9adb49fd1e/stardist/models/__init__.py#L20-L28).)


This should solve all the issues and offers the additional benefit of also adding your own custom trained models (if you have any) to the model registry.

Note that you don't have to use the pre-trained model registry, you could instead just download the models that interest you, put them in a shared folder, and just load them directly. E.g.

```python
from stardist.models import StarDist2D

model = StarDist2D(None, '2D_versatile_fluo', basedir='/path/to/shared/folder/')
...
```

Best,
Uwe

**Comment by DavidStirling** [comment]
Thanks Uwe, we did find a similar workaround but figured it was worth raising an issue anyway. There are likely to be circumstances where people want to use software which wraps packages like StarDIst without having to edit code they don't directly control.

**Comment by maweigert** [comment]
Hi @DavidStirling ,

thanks for the detailed issue! I agree that it would be nice to control the `cache_dir` (which seems not to be possible with keras itself). An alternative (even more crude) workaround could be to symlink the users `~/.keras/models/StarDist2D` to a common folder storing the models.

**Comment by DavidStirling** [comment]
> What do you mean by "having to edit code they don't directly control."?

While in this example you can use `StarDist2D`'s class directly, other packages may not support that flexibility and instead assume the default Keras cache directory by calling `get_model_folder` for everything. Changing that cache directory would therefore require either forking/editing the calling package or CSBDeep.
